lms/models/_mixins.py

```diff
@@ -13,3 +16,12 @@ class CreatedUpdatedMixin:
         onupdate=sa.func.now(),
         nullable=False,
     )
+
+
+class BulkUpsertMixin:
```

Comment:
Is this mixin here mostly as a documentation / cue of what's going on?

Reply:
Pretty much, there's much more detail in the longer comments above. The intention here is to:

 * Cause the parameters to be on the class (helps with auto complete)
 * Provide a positive declaration that this class supports being called with `bulk_upsert()`
 * Provide somewhere to attach documentation to give you a clue about what's going on

Otherwise these just end up being random parameters with no obvious connection to `bulk_upsert()`

Reply:
Don't you end up needing to add the attributes to the class anyway?

```
class File(BulkUpsertMixin):
    BULK_UPSERT_INDEX_ELEMENTS = ...
    BULK_UPSERT_UPDATE_ELEMENTS = ...
```

Won't autocomplete still pick up those class attrs, even without the mixin inheritance?

I think the existence of the `BULK_UPSERT_*` attrs is itself a positive assertion that the class supports bulk upsert too (duck typing).

And the documentation would go on the `CustomSession.bulk_upsert()` method?

I think I preferred the original mixin-only approach over the current custom session + mixin approach just because I'm not a fan of having a custom session object. But it does look to me like the mixin class is unnecessary if using either the custom session or the service approach.

Reply:
I think this is a good example of me attempting to address feedback and introduce compromise which results in worse outcomes. So don't think this is my preferred situation.

That said, I'll try and explain the difference. Adding a class attribute alone is a positive, but _obscure_ assertion. In the same way that referring to services with strings rather than class-names is a 1:1 correlation, but a much more obscure one. If you have a service called "user", then some _all_ of the places that ask for that service by name mean the service, and _all_ of the services implementing that name are the service, but only _some_ occurrences of "user" in the code base mean either. This means there is a connection, but you have to grep for it, and plenty of what you get back has nothing to do with it.

This is less likely to happen with a longer string like the one above, but the principal is the same.

I've noticed a couple of patterns in some of our discussions about related things:

 * A substitution of _discoverable_ for easily discovered
 * A reluctance to add bread crumbs to aid discovery
 
Whether something is easy to understand or obscure largely doesn't effect you once you understand it. I think a lot of our discussions often start from here rather than assuming you don't know how it works and are trying to find out. Lets assume you don't understand the mechanism and give an example of how you might discover it in each case. Each of the assumes you're starting from looking at the model and trying to work out what it can do to know that `bulk_upsert` is available, but you don't know it exists.

This describes one likely way to find this out, not all, or the only, or the most likely... (bum covering over)

### Solo mixin

* You look at the class
* You see the attributes and mixin
* The names relate, and you know mixins can add functionality _(leap)_
* You click through to the mixin and see the documentation and functionality
* You can use the functionality

### Mixin + functionality elsewhere (session / service)

* You look at the class
* You see the attributes and mixin
* The names relate, and you know mixins can add functionality _(leap)_
* You click through to the mixin and see the documentation about the functionality
* You follow the location
* You see the functionality
* You can use the functionality

### Attributes + functionality elsewhere

* You look at the class
* You see the attributes names and assume it means you can do something related to bulk upsetting _(leap)_
* You grep for the strings in the code base
* You ignore the results that don't relate to it and assume the one outside of the models implements related functionality _(leap)_
* You look at the code and see the documentation and functionality
* (In the case of string indexed services) You know services have a string to call them defined in a different file _(leap)_
* (In the case of string indexed services) Lookup the string name for the service
* You can use the functionality

All of these solutions work of course. It's not impossible, or perhaps even hard to work out any of them. But some of them require more mental leaps and knowledge to get to.